### PR TITLE
[Test Framework] Provide a way to specify addresses for the imports

### PR DESCRIPTION
**Pull request by SupunS**
work towards https://github.com/onflow/cadence/issues/331

## Description

The last piece of the puzzle 🙏 

One common pattern in Cadence projects is to define the imports as file locations and specify the addresses corresponding to each network at the [config file](https://developers.flow.com/tools/flow-cli/configuration#contracts). It would make life easier for devs, if they can write tests for such projects, without having to change the imports.

This PR introduces an API to set a similar configuration dynamically to the blockchain during tests. So the devs can specify the accounts/addresses to which each import maps, at a single place. So they don't have to manually update the imports, which can be a real hassle.

(I couldn't think of a better approach. Suggestions are most welcome!)

API:

```cadence
blockchain.useConfiguration(_ configs: Test.Configurations)
``` 

Sample Usage:

```cadence
pub var blockchain = Test.newEmulatorBlockchain()
pub var accounts: [Test.Account] = []

pub fun setup() {
    // Create accounts in the blockchain.

    let acct1 = blockchain.createAccount()
    accounts.append(acct1)

    let acct2 = blockchain.createAccount()
    accounts.append(acct2)

    // Set the configurations with the addresses

    blockchain.useConfiguration(Test.Configurations({
        "./FooContract": acct1.address,
        "./BarContract": acct2.address
    }))
}

```

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: test-framework/emulator_backend.go [hunk 1]
```diff
@@ -431,3 +451,30 @@ func newInterpreter() (*interpreter.Interpreter, error) {
...
+		if !ok {
+			continue
+		}
+
+		address := e.configurations.Addresses[location.String()]
+
+		locationStr := fmt.Sprintf(`"%s"`, location)
+		addressStr := fmt.Sprintf("0x%s", address)
+
+		code = strings.Replace(code, locationStr, addressStr, 1)
```
Comment (turbolent) [inline test-framework/emulator_backend.go:476]: Maybe use the offsets in the position info to replace the code for this particular import declaration only. The code may contain other string literals matching `locationStr` which should not be replaced.

**Review by turbolent — APPROVED** [review]
Nice!

File: test-framework/emulator_backend.go (hunk 1 above)
Comment (SupunS) [reply test-framework/emulator_backend.go]: Good idea. added with 22580ffec638e88bfb4c5d8e2edcf525cda4e4a9

File: runtime/stdlib/test.go [hunk 2]
```diff
@@ -1357,6 +1369,83 @@ func emulatorBackendDeployContractFunction(testFramework interpreter.TestFramewo
...
+			addresses, ok := configsValue.GetMember(
+				inter,
+				invocation.GetLocationRange,
+				addressesFieldName,
+			).(*interpreter.DictionaryValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			mapping := make(map[string]common.Address)
```
[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/stdlib/test.go]: added in 47e74ed43ef7b6749ee754e77bce4730db05aad6

File: runtime/stdlib/test.go [hunk 3]
```diff
@@ -1357,6 +1369,83 @@ func emulatorBackendDeployContractFunction(testFramework interpreter.TestFramewo
...
+
+const emulatorBackendUseConfigsFunctionDocString = `Use configurations function`
+
+var emulatorBackendUseConfigsFunctionType = func() *sema.FunctionType {
+	// The type of the 'useConfiguration' function of 'EmulatorBackend' (interface-implementation)
+	// is same as that of 'BlockchainBackend' interface.
+	typ, ok := blockchainBackendInterfaceType.Members.Get(emulatorBackendUseConfigsFunctionName)
+	if !ok {
+		panic(errors.NewUnexpectedError(
+			"cannot find type %s.%s",
```
[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/stdlib/test.go]: fixed in https://github.com/onflow/cadence/commit/47e74ed43ef7b6749ee754e77bce4730db05aad6

File: runtime/interpreter/test-framework.go [hunk 4]
```diff
@@ -76,6 +78,10 @@ type PublicKey struct {
 	SignAlgo  sema.SignatureAlgorithm
 }
 
+type Configurations struct {
```
[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/interpreter/test-framework.go:81]: fixed in https://github.com/onflow/cadence/commit/47e74ed43ef7b6749ee754e77bce4730db05aad6

**Pull request merged**